### PR TITLE
fix: refactor withThemeStyles

### PR DESCRIPTION
**Pull request by joshhowenstine**
## References

- Related Ticket(s): [LUI-704](https://ccp.sys.comcast.net/browse/LUI-704)

## Testing

Run the following commands to ensure a clean and up-to-date environment:

`
yarn clean && 
yarn install && yarn start
`

Test all stories and ensure everything operates as expected. Perform the following tests:
- All components should still function as expected.
- Components should update when global theme styles are changed in the Global Theme Values panel
- Specific component should update when tone is changed in the Component Theme Values  panel
- Components should be checked for memory leaks before this feature is released
- Allows for creation of custom accessors and mutators that map the props in the alias array (See changes in withUpdates)
- Allows for the creation of custom style aliases by setting `static get aliasStyles` on the component

### Theme update testing
1. Open up the Progress Bar Story in the application.
2. Right-click on the iframe containing the Progress Bar.
3. Open up Chrome Dev Tools.
4. Update the theme to see the result using `CONTEXT.setTheme({your theme updates})`.

#### Test the following theme updates and verify the expected results:

- `CONTEXT.setTheme({color: {fillNeutral: 0xFFF663399}})` - Should turn progress color purple.

- `CONTEXT.setTheme({color: {fillNeutral: ['#663399', 100]}})` - Should turn progress color purple.

- `CONTEXT.setTheme({color: {fillNeutral: 'theme.color.purple'}})` - Should turn progress color purple.

- `CONTEXT.setTheme({componentConfig: {ProgressBar: {tone: 'inverse'}}})` - Changing the value of tone from neutral, inverse, and brand should change the default tone of the ProgressBar across all instances of the Component.

- `CONTEXT.setTheme({componentConfig: {ProgressBar: {style: {progressColor: ['#663399', 100]}}}})` - Should ensure that progressColor is ALWAYS purple regardless of other color values or tone.

- `CONTEXT.setTheme({componentConfig: {ProgressBar: {style: {base:{h: 200}}}}})` - Will override the value for h set in ProgressBar.styles.js so it will be very tall.

- `CONTEXT.setTheme({componentConfig: {ProgressBar: {style: {tone: {inverse: {progressColor:  ['#0663399', 100]}}}}}})` - Only change progress color to purple when mode is set to inverse.

 - `CONTEXT.setTheme({componentConfig: {ProgressBar: {style: { mode: {focused: {h: 100}}}}}})` - Storybook automatically gives the ProgressBar focus so it will show as 100px height. If you change this to reflect unfocused the h will not be applied

- `CONTEXT.setTheme({componentConfig: {ProgressBar: {style: { mode: {focused: {tone: {brand: {h: 100}}}}}}}})` - Will only apply height of 100 when component has focus and is in brand mode

### CODE REVIEW

File: packages/@lightningjs/ui-components/src/components/Artwork/Artwork.test.js [hunk 1]
```diff
@@ -306,7 +306,7 @@ describe('Artwork', () => {
     };
     await artwork.__updateSpyPromise;
     expect(artwork._FillColor).not.toBeUndefined();
-    expect(artwork._FillColor.color).toBe(0xfff663399);
+    expect(artwork._FillColor.color).toBe(3004700697); // Equal to 0xfff663399 decimal format
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/components/Artwork/Artwork.test.js]: does this change matter? I feel like the previous is easier to parse and know what color is being used

File: packages/@lightningjs/ui-components/src/components/Button/Button.stories.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ import lightning from '../../assets/images/ic_lightning_white_32.png';
 import { createModeControl } from '../../docs/utils';
 import { CATEGORIES } from '../../docs/constants';
 import { controlDescriptions } from '../../docs/constants';
-
+import context from '../../globals/context'
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/components/Button/Button.stories.js]: remove

File: packages/@lightningjs/ui-components/src/mixins/withExtensions/index.js [hunk 3]
```diff
@@ -31,9 +31,10 @@ export default function withExtensions(Base) {
 
     static get __componentName() {
       if (super.__componentName) return super.__componentName;
-      throw new Error(
-        `A valid static __componentName property is required for theming to work properly. Please add this to the ${this.constructor.name} class.`
-      );
+      if (['Base', '_class'].includes(super.name)) return 
+      // throw new Error(
+      //   `A valid static __componentName property is required for theming to work properly. Please add this to the ${this.constructor.name} class.`
+      // );
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withExtensions/index.js]: uncomment

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test-skip.js [hunk 4]
```diff
@@ -0,0 +1,314 @@
+import StyleManager from './StyleManager';
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test-skip.js]: Rename file so these run?

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/theme-utils.test-skip.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
+﻿import {
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/theme-utils.test-skip.js]: rename file so these run?

File: scripts/generate-style-json-schema.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
+/* eslint-disable no-restricted-syntax */
+
+const fs = require('fs');
```
Comment (erautenberg) [inline scripts/generate-style-json-schema.js]: is this at all connected to the SUITE stuff? Would it be better to wait until that work is finished before introducing any of the validation here as it is out of the scope of withThemeStyles?

File: yarn.lock [hunk 7]
```diff
@@ -2629,7 +2629,7 @@ __metadata:
   dependencies:
     "@babel/core": ^7.19.3
     "@lightningjs/ui-components-test-utils": "workspace:^"
-    "@lightningjs/ui-components-theme-base": "workspace:^"
+    "@lightningjs/ui-components-theme-base": 1.0.2
```
Comment (erautenberg) [inline yarn.lock]: why the change?

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/utils.js [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+
+  return finalStyle;
+};
+
+/**
+ * Traverse up the prototype chain to create an array of all the styles that are present in the Components ancestors
+ * @param {object} componentObj - The component object to get the style chain from.
+ * @returns {{ style: (object | function) }[]} - An array of style objects containing either an object of styles or a function to return an object of styles.
+ */
+export function getStyleChain(componentObj) {
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/utils.js]: if you're going to set all the other functions as variables, might as well do it here too for consistency

File: packages/@lightningjs/ui-components/package.json [hunk 9]
```diff
@@ -45,7 +45,7 @@
     "semantic-release-monorepo": "^7.0.5"
   },
   "dependencies": {
-    "@lightningjs/ui-components-theme-base": "workspace:^",
+    "@lightningjs/ui-components-theme-base": "1.0.2",
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/package.json]: why the change?

File: packages/@lightningjs/ui-components/src/components/Artwork/Artwork.test.js (hunk 1 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/components/Artwork/Artwork.test.js]: With the new withThemeStyles process Javascript is interpreting the hexadecimal value as a decimal number. I am not 100% sure why this test had to change to pass to be honest.

File: packages/@lightningjs/ui-components/src/mixins/withExtensions/index.js (hunk 3 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withExtensions/index.js]: I'm not 100% sure this is also implemented properly. Perhaps you could help me vet the approach of throwing an error here.

File: scripts/generate-style-json-schema.js (hunk 6 above)
Comment (joshhowenstine) [reply scripts/generate-style-json-schema.js]: Ooops. Im not sure how that snuck in. Removed.

File: yarn.lock (hunk 7 above)
Comment (joshhowenstine) [reply yarn.lock]: It was part of testing between projects. Change reverted.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/utils.js (hunk 8 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withThemeStyles/utils.js]: Changed to match

File: packages/@lightningjs/ui-components/package.json (hunk 9 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/package.json]: Reverted change. Was used for testing

File: packages/@lightningjs/ui-components-test-utils/src/lng-test-env.js [hunk 10]
```diff
@@ -30,7 +30,7 @@ const extractWidthHeight = src => {
 
 export default class LightningUIEnvironment extends JSDOMEnvironment {
   async setup() {
-    super.setup();
+    await super.setup();
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components-test-utils/src/lng-test-env.js]: Can you add comments about why this is necessary?

File: packages/@lightningjs/ui-components/src/components/TextBox/TextBox.test.js [hunk 11]
```diff
@@ -142,7 +142,8 @@ describe('TextBox', () => {
       );
     });
 
-    it('should set text styles to the defaultTextStyle if textStyle is not provided', async () => {
+    // TODO: Check with team to see if defaultTextStyle should still be suppported
+    it.skip('should set text styles to the defaultTextStyle if textStyle is not provided', async () => {
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/components/TextBox/TextBox.test.js]: can be removed, but need to also remove this initial check in TextBox.js
```js
  get _textStyleSet() {
    const fontStyle = {
      ...(this.theme.typography[this.style.defaultTextStyle] ||
        this.theme.typography.body1),
```

File: packages/@lightningjs/ui-components/src/components/Tile/Tile.stories.js [hunk 12]
```diff
@@ -48,10 +48,10 @@ export const Tile = () =>
             ratioX: 16,
             ratioY: 9,
             upCount: 3
-          },
-          artwork: {
-            src: 'https://image.tmdb.org/t/p/w500/zHdQ6yaqDf3OQO5uhr0auAgwK6O.jpg'
           }
+          // artwork: {
+          //   src: 'https://image.tmdb.org/t/p/w500/zHdQ6yaqDf3OQO5uhr0auAgwK6O.jpg'
+          // }
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/components/Tile/Tile.stories.js]: Uncomment?

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+export default class StyleManager extends lng.EventEmitter {
+  /**
+   * Create a Style Manager instance.
+   * @param {Object} options - The options object.
+   * @param {Object} options.component - The LightningJS component instance.
+   */
+  constructor({ component = {} } = {}) {
+    super(...arguments);
+    this.resolving = false;
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: this doesn't seem to be used anywhere, can it be removed?

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+      window.LUI_STYLE_CACHE = new Map();
+    }
+
+    this.component = component;
+    this._style = {};
+    // Initial Update
+    this._update();
+
+    if (process?.env?.NODE_ENV === 'test') {
+      this.update = this._update; // Avoid race conditions in tests
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: can you clarify this at all? It looks like `this.update()` is only ever called here in this file. Why not just change it to the underscore value and remove this? Is this part of lng.EventEmitter?

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    return window.LUI_STYLE_CACHE.get(key);
+  }
+
+  /**
+   * Updates the component's style based on changes in tone, mode, theme, and `_componentLevelStyle`.
+   * Clears the style cache if tone or mode have changed, and clears the source cache if `_componentLevelStyle` has changed.
+   * Generates the component's style source and style objects, saves them to the cache, and emits a `styleUpdate` event with the updated style.
+   * @returns {Promise<void>}
+   */
+  async _update() {
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: why does this need to be async?

Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: Was trying to keep it async incase we ever found value in processing styles with web workers in the future.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js (hunk 14 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: update in place of _update only was put in place to make sure the call the _update is debounced. I could rename it if you think it is more clear however.

**Comment by ImCoolNowRight** [comment]
Yes, I agree @ImCoolNowRight . I would also prefer that anything that is not base, mode, tone in style to be able to overwrite everything. But after some conversation about alignment with the other teams and themes it was decided that this unfortunatly would be an anti-pattern. 


> CONTEXT.setTheme({componentConfig: {ProgressBar: {style: {progressColor: ['#663399', 100]}}}}) - Should ensure that progressColor is ALWAYS purple regardless of other color values or tone.

This feels **_very_** weird to me. Maybe it isn't such a big deal in this particular example, but I'm just thinking about how it might work with some other components. For example, if someone wants to change which text color is being used by default  for the Button (i.e. in the `base` of the styles), this would force them to have to redefine the mode and tone setups if they still want them, right?

**Comment by ImCoolNowRight** [comment]
> > CONTEXT.setTheme({componentConfig: {ProgressBar: {style: {progressColor: ['#663399', 100]}}}}) - Should ensure that progressColor is ALWAYS purple regardless of other color values or tone.
> 
> This feels **_very_** weird to me. Maybe it isn't such a big deal in this particular example, but I'm just thinking about how it might work with some other components. For example, if someone wants to change which text color is being used by default for the Button (i.e. in the `base` of the styles), this would force them to have to redefine the mode and tone setups if they still want them, right?

Ok, I just read the next line, and I think that if the `progressColor` was nested inside of a `base` object, it achieves what I was concerned about. If that's the case, you can disregard the above.

**Comment by ImCoolNowRight** [comment]
Ran through the code and all of your steps to test. Other than Emily's comments and the lint error, everything looks good and seems to be working as expected.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js (hunk 14 above)
Comment (anthony9187) [reply packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: i think naming it something else would be better, also should it be underscored or is this part of the API?

File: packages/@lightningjs/ui-components/src/components/Key/Key.stories.js [hunk 16]
```diff
@@ -32,42 +32,42 @@ export default {
...
+//     }
+//   },
+//   title: {
+//     control: 'text',
+//     description: 'Key character',
+//     table: {
+//       defaultValue: { summary: 'undefined' }
+//     }
+//   }
+// };
```
Comment (ImCoolNowRight) [inline packages/@lightningjs/ui-components/src/components/Key/Key.stories.js]: Is this still in-progress, can it be removed, or is it to circle back to in a separate PR?

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+ */
+
+import {
+  generateComponentStyleSource,
+  generateStyle,
+  replaceAliasValues
+} from './utils.js';
+import { context } from '../../globals/index.js';
+import { debounce } from '../../utils/index.js';
+import cache from './cache.js'
```
Comment (ImCoolNowRight) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/StyleManager.js]: Just calling out the lint error here.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/cache.js [hunk 18]
```diff
@@ -0,0 +1,3 @@
+const themeStyleCache = new Map();
+
+export default themeStyleCache;
```
Comment (ImCoolNowRight) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/cache.js]: Calling out another lint error.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js [hunk 19]
```diff
@@ -0,0 +1,326 @@
+import StyleManager from './StyleManager';
+import lng from '@lightningjs/core';
+import { jest } from '@jest/globals';
+import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';
+import context from '../../globals/context';
+import cache from './cache'
```
Comment (ImCoolNowRight) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js]: One more.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+  }
+);
+
+describe('StyleManager', () => {
+  let component;
+  let styleManager;
+
+  beforeEach(() => {
+    console.log('CACHE', cache)
+    cache.clear()
```
Comment (ImCoolNowRight) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js]: Ok, I'll stop now. You can skim through on your own 😬

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+      const themeCache = styleManager._getCache(name);
+      expect(themeCache).toEqual({
+        ids: [component.__id],
+        payload
+      });
+    });
+  });
+
+  describe('_update', () => {
+    it.skip('should update the component style based on changes in tone, mode, and _componentLevelStyle', () => {
```
Comment (ImCoolNowRight) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js]: If these are going to stay skipped, is it worth throwing in a comment for these about why they're being skipped so that we know where to pick up with them eventually?

**Review by arwehrman — COMMENTED** [review]
On Card ( or any variation of card) it doesn't appear that changing tone in Component Style Theme Panel works as expected. When changing from neutral to inverse the background color should change.

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/index.js [hunk 22]
```diff
@@ -131,6 +131,7 @@ export default function withThemeStyles(Base, mixinStyle = {}) {
      */
     _updateThemeComponent() {
       if (!this.style) return;
+      if (!this._isAttached()) return;
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/index.js:148]: maybe just combine this with the above line? 🤷‍♀️

File: packages/@lightningjs/ui-components/src/components/Icon/Icon.js [hunk 23]
```diff
@@ -165,8 +165,6 @@ export default class Icon extends Base {
     return {
       texture,
       shader,
-      w: this.w,
-      h: this.h,
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/components/Icon/Icon.js:169]: why were these removed?

File: packages/@lightningjs/ui-components/src/components/Provider/Provider.js [hunk 24]
```diff
@@ -67,12 +67,17 @@ export default class Provider extends Base {
...
         centerInParent: true,
         radius: this.disableRadius ? 0 : this.style.radius,
-        alpha: this.style.alpha,
-        style: provider.style
+        alpha: this.style.alpha
       };
+
+      if (provider.style) {
+        patch.style = provider.style;
+      }
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/components/Provider/Provider.js]: does this matter? if style is set to undefined, does anything actually happen? Or could we keep it inline and do provider.style || {}?

File: packages/@lightningjs/ui-components/src/mixins/withLayout/index.js [hunk 25]
```diff
@@ -102,7 +102,7 @@ export default function withLayout(Base) {
             : calculatedWidth; // Width must be set first in order for Cards to be displayed properly
         this.h = calculatedHeight;
         if (this._itemLayout && this._itemLayout.circle && this.style) {
-          if (-1 < this._componentStyleSourceProps.indexOf('radius')) {
+          if (-1 < this.style?.radius) {
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withLayout/index.js]: what's the -1 for in this context? For the above, I think it makes sense as its checking if it exist

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js [hunk 26]
```diff
@@ -0,0 +1,325 @@
+import StyleManager from './StyleManager';
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js]: can we rename this file to match "StyleManager" instead of "style-manager"

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/withThemeStyles.test.js [hunk 27]
```diff
@@ -0,0 +1,4 @@
+describe('withThemeStyles', () => {
+  beforeEach(() => {});
+  it.skip('should construct with default behavior', () => {});
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/withThemeStyles.test.js]: is this intentional or does this need to be fleshed out in a future ticket? (or write a todo here?)

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/index.js [hunk 28]
```diff
@@ -16,417 +16,166 @@
...
-          this[`_${prop}`] = this._componentConfig[prop];
-        }
+      this._hSetByUser = false;
+      this._styleManager = new StyleManager({ component: this });
+      this._style = this._styleManager.style; // Set the style for the first time. After this is will be updated by events
+      this._styleManager.on('styleUpdate', () => {
+        this._style = this._styleManager.style;
+        this.queueThemeUpdate();
       });
+      this._wSetByUser = false;
```
Comment (erautenberg) [inline packages/@lightningjs/ui-components/src/mixins/withThemeStyles/index.js]: Being picky, but can this be next to the `hSetByUser` variable above?

File: packages/@lightningjs/ui-components/src/components/Icon/Icon.js (hunk 23 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/components/Icon/Icon.js:169]: If these are left in place there is an infinite loop in badge icon with pngs

File: packages/@lightningjs/ui-components/src/mixins/withLayout/index.js (hunk 25 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withLayout/index.js]: Fixed

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/withThemeStyles.test.js (hunk 27 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withThemeStyles/withThemeStyles.test.js]: Future ticket. Added TODO

**Review by erautenberg — APPROVED** [review]

File: packages/@lightningjs/ui-components/src/components/Icon/Icon.js (hunk 23 above)
Comment (erautenberg) [reply packages/@lightningjs/ui-components/src/components/Icon/Icon.js:169]: Coooool

**Review by ImCoolNowRight — APPROVED** [review]

File: packages/@lightningjs/ui-components/src/components/Key/Key.stories.js (hunk 16 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/components/Key/Key.stories.js]: Fix, thanks!

File: packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js (hunk 21 above)
Comment (joshhowenstine) [reply packages/@lightningjs/ui-components/src/mixins/withThemeStyles/style-manager.test.js]: Removing these tests for now. I will have a fast follow PR to add more in depth unit tests for this feature.

**Comment by joshhowenstine** [comment]
> On Card ( or any variation of card) it doesn't appear that changing tone in Component Style Theme Panel works as expected. When changing from neutral to inverse the background color should change.

Thanks @arwehrman ! Should be good now
